spec/services/idv/gpo_mail_spec.rb

```diff
@@ -3,76 +3,181 @@
+
+      context 'but MAX_MAIL_EVENTS is non-zero' do
+        context 'and MAIL_EVENTS_WINDOW is zero' do
+          let(:mail_events_window_days) { 0 }
 
-        expect(subject.mail_spammed?).to eq false
+          context 'and MINIMUM_WAIT_BEFORE_ANOTHER_USPS_LETTER is zero' do
+            let(:minimum_wait_before_another_usps_letter_in_hours) { 0 }
+
+            it 'returns false' do
+              expect(subject.mail_spammed?).to be_falsey
+            end
+          end
+
+          context 'and MINIMUM_WAIT_BEFORE_ANOTHER_USPS_LETTER is non-zero' do
+            it 'returns true' do
+              expect(subject.mail_spammed?).to eq true
+            end
+          end
+        end
+
+        context 'and MAIL_EVENTS_WINDOW is non-zero' do
+          context 'and MINIMUM_WAIT_BEFORE_ANOTHER_USPS_LETTER is zero' do
+            let(:minimum_wait_before_another_usps_letter_in_hours) { 0 }
+
+            it 'returns true' do
+              expect(subject.mail_spammed?).to eq true
+            end
+          end
+
+          context 'and MINIMUM_WAIT_BEFORE_ANOTHER_USPS_LETTER is non-zero' do
+            it 'returns true' do
+              expect(subject.mail_spammed?).to eq true
+            end
+          end
+        end
       end
     end
-  end
 
-  def event_create(hash)
-    uuid = 'foo'
-    remote_ip = '127.0.0.1'
-    user = hash[:user]
-    event = hash[:event_type]
-    now = Time.zone.now
-    updated_at = hash[:updated_at] || now
-    device = Device.find_by(user_id: user.id, cookie_uuid: uuid)
-    if device
-      device.last_used_at = now
-      device.last_ip = remote_ip
-      device.save
-    else
-      last_login_at = Time.zone.now
-      device = Device.create(
-        user_id: user.id,
-        user_agent: '',
-        cookie_uuid: uuid,
-        last_used_at: last_login_at,
-        last_ip: remote_ip,
-      )
+    context 'when a user has a recent GPO request' do
+      let(:user) do
+        user = create(:user, :deactivated_password_reset_profile, :with_pending_gpo_profile)
+
+        # at this point, the gpo is attached to the pending profile. Move it.
+        not_pending_profile = user.profiles.where.not(id: user.pending_profile.id).first
+        GpoConfirmationCode.first.update(profile: not_pending_profile)
+
+        user
+      end
+
+      context 'that is not attached to their pending profile' do
+        it 'returns false' do
+          expect(subject.mail_spammed?).to be_falsey
+        end
+      end
+
+      context 'and no pending profile' do
+        before do
+          user.send(:instance_variable_set, :@pending_profile, nil)
+        end
+
+        it 'returns false' do
+          expect(subject.mail_spammed?).to be_falsey
+        end
+      end
     end
-    Event.create(
-      user_id: user.id,
-      device_id: device.id,
-      ip: remote_ip,
-      event_type: event,
-      created_at: updated_at, updated_at: updated_at
+  end
+
+  def enqueue_gpo_letter_for(user, at: Time.zone.now)
```

Comment:
What a useful method! We might want to make it available in a helper, although looking around I didn't see an obvious place for it.

I wonder if it would be better to travel to `at` and create the profile & confirmation code, rather than updating the timestamps. I could see that creating issues later if the profile timestamps are after the confirmation code timestamps.

Directly setting `@pending_profile` could also cause trouble in the long run. Is there a reason not to call user.pending_profile? This writes the profile to the db, so the call should work.

Reply:
I tend to prefer explicitly setting up the required data structures for unit tests, such as this one. That way, the tests make it clear exactly what plumbing the code under test depends on. For feature tests, `travel_to` is absolutely the best way to go.

Reply:
Sure, let me know if you want to pair. Maybe we can look into not restarting idv so many times, too.

Reply:
Revised to avoid setting the instance variable directly; worked it out.